### PR TITLE
feat(turbo): add platform env support

### PR DESCRIPTION
**Pull request by tknickman**
### Description

Support reading what variables are set by a user in CI, and diffing those with those in the execution environment. 

## Testing
```
TURBO_SYSTEM_ENV=THIS_NEAT_ENV,SUPER_COOL,TEST THIS_NEAT_ENV=1 ./target/debug/turbo build --skip-infer --force --cwd examples/basic
```

### CODE REVIEW

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 1]
```diff
@@ -881,6 +924,18 @@ impl ExecContext {
...
 
+        if !self.task_cache.is_caching_disabled() {
+            let missing_platform_env = self.platform_env.validate(&self.execution_env);
+            if !missing_platform_env.is_empty() {
+                let _ = self.warnings.lock().map(|mut warnings| {
+                    warnings.push(TaskWarning {
+                        task_id: self.task_id_for_display.clone(),
+                        missing_platform_env,
+                    });
+                });
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:935]: It's probably worth panicking here if locking fails. Locking should only fail if another thread panicked while holding the lock.
```suggestion
                self.warnings.lock().expect("warnings lock poisoned").push(TaskWarning {
                        task_id: self.task_id_for_display.clone(),
                        missing_platform_env,
                 });|
```

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 2]
```diff
@@ -381,6 +384,35 @@ impl<'a> Visitor<'a> {
 
         let global_hash_summary = GlobalHashSummary::try_from(global_hash_inputs)?;
 
+        // output any warnings that we collected while running tasks
+        if let Ok(warnings) = self.warnings.lock() {
+            if !warnings.is_empty() {
+                println!();
+                warn!("finished with warnings");
+                println!();
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:392]: I think we want to print all of this to `stderr`
```suggestion
                eprintln!();
                warn!("finished with warnings");
                eprintln!();
```

File: crates/turborepo-env/src/platform.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        missing: Vec<String>,
+        task_id_for_display: &str,
+        color_config: ColorConfig,
+    ) {
+        cprintln!(color_config, YELLOW, "{}", task_id_for_display);
+        for key in missing {
+            cprint!(color_config, GREY, "  - ");
+            cprint!(color_config, GREY, "{}\n", key);
+        }
+        println!();
```
Comment (chris-olszewski) [inline crates/turborepo-env/src/platform.rs:102]: These are all printing to `stdout` instead of `stderr`

**Review by chris-olszewski — APPROVED** [review]
This is awesome! Only thing that needs changing is moving all the outputs to `stderr` to avoid breaking uses of `turbo --dry=json` that might happen in CI.

File: crates/turborepo-lib/src/task_graph/visitor.rs (hunk 2 above)
Comment (tknickman) [reply crates/turborepo-lib/src/task_graph/visitor.rs:392]: Good call

File: crates/turborepo-env/src/platform.rs (hunk 3 above)
Comment (tknickman) [reply crates/turborepo-env/src/platform.rs]: Fixed!

**Pull request merged**